pyro/distributions/one_two_matching.py

```diff
@@ -0,0 +1,184 @@
+    number of belief propagation iterations via the ``bp_iters`` argument.  The
+    :meth:`log_partition_function` and :meth:`log_prob` methods use a Bethe
+    approximation [1,2,3] with fractional free energy [4].
+
+    **References:**
+
+    [1] Michael Chertkov, Lukas Kroc, Massimo Vergassola (2008)
+        "Belief propagation and beyond for particle tracking"
+        https://arxiv.org/pdf/0806.1199.pdf
+    [2] Bert Huang, Tony Jebara (2009)
+        "Approximating the Permanent with Belief Propagation"
+        https://arxiv.org/pdf/0908.1769.pdf
+    [3] Pascal O. Vontobel (2012)
+        "The Bethe Permanent of a Non-Negative Matrix"
+        https://arxiv.org/pdf/1107.4196.pdf
+    [4] M Chertkov, AB Yedidia (2013)
+        "Approximating the permanent with fractional belief propagation"
+        http://www.jmlr.org/papers/volume14/chertkov13a/chertkov13a.pdf
+
+    :param Tensor logits: An ``(2 * N, N)``-shaped tensor of edge logits.
+    :param int bp_iters: Optional number of belief propagation iterations. If
+        unspecified or ``None`` expensive exact algorithms will be used.
+    """
+    arg_constraints = {"logits": constraints.real}
+    has_enumerate_support = True
+
+    def __init__(self, logits, *, bp_iters=None, validate_args=None):
+        if logits.dim() != 2:
+            raise NotImplementedError("OneTwoMatching does not support batching")
+        assert bp_iters is None or isinstance(bp_iters, int) and bp_iters > 0
+        self.num_sources, self.num_destins = logits.shape
+        assert self.num_sources == 2 * self.num_destins
+        self.logits = logits
+        batch_shape = ()
+        event_shape = (self.num_sources,)
+        super().__init__(batch_shape, event_shape, validate_args=validate_args)
+        self.bp_iters = bp_iters
+
+    @constraints.dependent_property
+    def support(self):
+        return OneTwoMatchingConstraint(self.num_destins)
+
+    @lazy_property
+    def log_partition_function(self):
+        if self.bp_iters is None:
+            # Brute force.
+            d = self.enumerate_support()
+            s = torch.arange(d.size(-1), dtype=d.dtype, device=d.device)
+            return self.logits[s, d].sum(-1).logsumexp(-1)
+
+        # Approximate mean field beliefs b via Sinkhorn iteration.
+        # We find that Sinkhorn iteration is more robust and faster than the
+        # optimal belief propagation updates suggested in [1-4].
+        finfo = torch.finfo(self.logits.dtype)
+        shift = self.logits.data.max(1, True).values
+        shift.clamp_(min=finfo.min, max=finfo.max)
+        p = (self.logits - shift).exp().clamp(min=finfo.tiny ** 0.5)
+        d = 2 / p.sum(0)
+        for _ in range(self.bp_iters):
+            s = 1 / (p @ d)
+            d = 2 / (s @ p)
+        b = s[:, None] * d * p
+
+        # Evaluate the Bethe free energy, adapting [4] Eqn 4 to one-two
+        # matchings.
+        b = b.clamp(min=finfo.tiny ** 0.5)
```

Comment:
isn't `** 0.5` a bit strange?

Reply:
not sure why `finfo.tiny ** 0.5` is strange? I wanted to clamp to something small such that its square is still nonzero. So `sqrt(finfo.tiny)` seemed like a reasonable choice. Also this avoided NAN grads in my model, whereas both the smaller `finfo.tiny` and the larger `finfo.eps` failed with NANs. Also I prefer to write `math.sqrt(x)` as `x ** 0.5` because it is shorter and doesn't require an import.